src/Common/CSharp/Analysis/ReduceIfNesting/IfLocalVariableAnalysis.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Josef Pihrt and Contributors. Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Roslynator.CSharp.Analysis.ReduceIfNesting;
+
+internal static class IfStatementLocalVariableAnalysis
+{
+    public static bool DoDeclaredVariablesOverlapWithOuterScope(
+        IfStatementSyntax ifStatement,
+        SemanticModel semanticModel
+    )
+    {
+        
+        var ifVariablesDeclared = semanticModel.AnalyzeDataFlow(ifStatement)!
+            .VariablesDeclared;
+        
+        if (ifVariablesDeclared.IsEmpty)
+            return false;
+        
+        var outerScope = GetOuterScope(ifStatement);
+        
+        if (outerScope == null)
+            return true;
+
+        IList<ISymbol> parentVariablesDeclared;
+        if (outerScope is SwitchSectionSyntax switchSection)
+        {
+            parentVariablesDeclared = switchSection.Statements.SelectMany(s => semanticModel.AnalyzeDataFlow(s)!.VariablesDeclared).ToList();
```

Comment:
By lazy initialization I meant following:

```
List<ISymbol> allDeclaredVariables = null;
foreach (StatementSyntax statement in switchSection.Statements)
{
    ImmutableArray<ISymbol> variables = semanticModel.AnalyzeDataFlow(statement)!.VariablesDeclared;

    if (variables.Any())
        (allDeclaredVariables ??= new List<ISymbol>()).AddRange(variables);
}

parentVariablesDeclared = allDeclaredVariables;
```

Reply:
I took a look at this part of the code and I think that parent scope will be always block or switch section. If the parent scope is, let's say, foreach statement then it means that the if statement is actually an embedded statement (and current implementation will return if statement itself as a parent scope). And if statement's parent will never be method, constructor etc.

So I propose following changes that would simplify the code:

(`StatementListInfo` is basically something that contains a list of statements, i.e. block or switch section)

```cs
foreach (StatementSyntax statement in SyntaxInfo.StatementListInfo(ifStatement).Statements)
{
    if (statement == ifStatement)
        continue;

    foreach (ISymbol parentVariable in semanticModel.AnalyzeDataFlow(statement)!.VariablesDeclared)
    {
        foreach (ISymbol ifVariable in ifVariablesDeclared)
        {
            if (ifVariable.Name == parentVariable.Name)
                return true;
        }
    }
}
```

Reply:
Yes I think that makes sense.